utopia-remix/app/components/sharingDialog.tsx

```diff
@@ -234,79 +223,178 @@ const OwnerCollaboratorRow = React.memo(() => {
 
   const isCollaborative = React.useMemo(() => {
     return projectAccessLevel === AccessLevel.COLLABORATIVE
   }, [projectAccessLevel])
 
   return accessRequests
     .sort((a, b) => {
-      if (a.status !== b.status) {
-        return a.status - b.status
-      }
-      return moment(a.updated_at).unix() - moment(b.updated_at).unix()
+      return moment(a.created_at).unix() - moment(b.created_at).unix()
     })
     .map((request) => {
       const user = request.User
       if (user == null) {
         return null
       }
 
-      const status = request.status
-      const canBeApproved = isCollaborative && status === AccessRequestStatus.PENDING
-      const color =
-        status === AccessRequestStatus.PENDING
-          ? 'gray'
-          : status === AccessRequestStatus.APPROVED
-          ? 'green'
-          : 'red'
+      const status = mustAccessRequestStatus(request.status)
+
       return (
         <CollaboratorRow
-          key={request.token}
+          key={`collaborator-${request.token}`}
           picture={user.picture}
           name={user.name ?? user.email ?? user.user_id}
           isDisabled={!isCollaborative}
         >
-          {canBeApproved ? (
-            <Button size='1' variant='ghost' onClick={onApprove(request.token)}>
-              Approve
-            </Button>
-          ) : (
+          {when(
+            isCollaborative,
+            <AccessRequestDropdown
+              status={request.status}
+              onApprove={onUpdateAccessRequest(request.token, 'approve')}
+              onReject={onUpdateAccessRequest(request.token, 'reject')}
+              onDestroy={onUpdateAccessRequest(request.token, 'destroy')}
+            />,
+          )}
+          {unless(
+            isCollaborative,
             <Text
               size='1'
-              color={color}
               style={{
-                fontStyle: !isCollaborative ? 'italic' : 'normal',
                 cursor: 'default',
+                fontStyle: !isCollaborative ? 'italic' : 'normal',
               }}
             >
               {when(status === AccessRequestStatus.PENDING, 'Pending')}
-              {when(status === AccessRequestStatus.APPROVED, 'Approved')}
-              {when(status === AccessRequestStatus.REJECTED, 'Rejected')}
-            </Text>
+              {when(status === AccessRequestStatus.APPROVED, 'Collaborator')}
+              {when(status === AccessRequestStatus.REJECTED, 'Blocked')}
+            </Text>,
           )}
         </CollaboratorRow>
       )
     })
 }
 
+const AccessRequestDropdown = React.memo(
+  ({
+    status,
+    onApprove,
+    onReject,
+    onDestroy,
+  }: {
+    status: AccessRequestStatus
+    onApprove: () => void
+    onReject: () => void
+    onDestroy: () => void
+  }) => {
+    return (
+      <DropdownMenu.Root>
+        <DropdownMenu.Trigger>
+          {/* this needs to be inlined (and as a ternary) because DropdownMenu.Trigger requires a direct single child */}
```

Comment:
Maybe we should pull it out into its own component in that case?

Reply:
yeah my comment is meant for that case too: it needs an _explicit_ inline child, because otherwise it will explode due to the component not accepting refs (?!). You can see I tried doing it here but had to revert it back :( [`645b73a` (#5155)](https://github.com/concrete-utopia/utopia/pull/5155/commits/645b73a6fd7aec0c9439d4ac1108af435dd6dc1f#diff-76cf853dcdab751a5e1d17fd9e05bb992d915ef606780a5203cb86bc372262c2R379)